docs/architecture/adr-064-batch-verification.md

```diff
@@ -0,0 +1,85 @@
+# ADR 064: Batch Verification
+
+## Changelog
+
+- January 28, 2021: Created (@marbar3778)
+
+## Context
+
+Tendermint uses public private key cryptography for validator signing. When a block is proposed and voted on validators sign a message representing acceptance or rejection of a block. These signatures are also used to verify previous blocks are correct if a node is syncing. Currently, Tendermint requires each signature to be verified individually, this leads to a slow down of block times.
+
+## Alternative Approaches
+
+- Signature aggregation
+  - Signature aggregation is an alternative to batch verification. Signature aggregation leads to fast verification and smaller block sizes. At the time of writing this ADR there is on going work to enable signature aggregation in Tendermint. The reason why we have opted to not introduce it at this time is because every validator signs a unique message. One of the two benefits can be done, aggregation, but verification would suffer because the software would need to verify each signature individually due to their uniqueness. For example if we were to implement signature aggregation with BLS, there could be a potential slow down of 10x-100x in verification speeds.
+
+## Decision
+
+Adopt Batch Verification.
+
+## Detailed Design
+
+A new interface will be introduced. This interface will have three methods `NewBatchVerifier`, `Add` and `VerifyBatch`.
+
+```go
+type BatchVerifier interface {
+  NewBatchVerifier() BatchVerifier // NewBatchVerifier create a new verifier where keys, signatures and messages can be added as entries
```

Comment:
Why have the constructor be part of the interface? i.e. an interface returns itself in the constructor. IMHO we can remove this from the interface.

Reply:
The reason I added it was in the code we will need to call `NewBatchVerifier` and multiple key types may support batch verification. Can we assume they will all have the same constructor if its not part of the interface?

Reply:
Yeah, we'd have a constructor for each key type. No need to have it be part of the interface, otherwise construction would look weird.